File: fp_py/src/main/fruitpunch/native/physics/body.py
from fruitpunch.native.elements import Element
import libfp_core
from fruitpunch.native.native_class import NativeClass
from fruitpunch.native.graphics.vertex import Point

class BodyType:
    static = libfp_core.BodyType.static
    dynamic = libfp_core.BodyType.dynamic
    kinematic = libfp_core.BodyType.kinematic

class Body(Element):
    '''
    Represents a box2d body and inherits Element.     
    
    Any element that is a transformable and has a body added to it will 
    automatically be transformed into a body that behaves physically on 
    the screen by the means of Shapers registered into the world class.     
    '''

    def __init__(self, bodytype=None):
        NativeClass.__init__(self)
        
        self.store(libfp_core.Body())
        
        if not bodytype is None:
            self.type = bodytype        

    # ---------------------------------------------------------------------
    # Native Methods
    # ---------------------------------------------------------------------

    def create(self):
        '''
        Creates the body fixture and shape based on its parent.
        
        '''

        return self.run_native("create")
    # _____________________________________________________________________

    def render_frame(self):
        '''
        Updates the parent with the position and rotation of the body in 
        the physics world.
        '''

        return self.run_native("render_frame")
    # _____________________________________________________________________

    def get_transformable(self):
        return self.run_native("get_transformable")
    # _____________________________________________________________________

    def set_transformable(self, transformable):
        return self.run_native("set_transformable", transformable)
    # _____________________________________________________________________

    def get_world(self):
        return self.run_native("get_world")
    # _____________________________________________________________________

    def set_world(self, world):
        return self.run_native("set_world", world)
    # _____________________________________________________________________

    def get_type(self):
        return self.run_native("get_type")
    # _____________________________________________________________________

    def set_type(self, btype):
        return self.run_native("set_type", btype)
    # _____________________________________________________________________

    def get_shape_scale(self):
        return self.run_native("get_shape_scale")
    # _____________________________________________________________________

    def set_shape_scale(self, shapeScale):
        return self.run_native("set_shape_scale", shapeScale)
    # _____________________________________________________________________

    def get_linear_velocity(self):
        return Point.create_attached(self._native.get_linear_velocity, self._native.set_linear_velocity)
    # _____________________________________________________________________

    def set_linear_velocity(self, velocity):
        p = libfp_core.Point()
        p.x = velocity.x
        p.y = velocity.y
        
        return self.run_native("set_linear_velocity", p)
    # _____________________________________________________________________

    def get_angular_velocity(self):
        return self.run_native("get_angular_velocity")
    # _____________________________________________________________________

    def set_angular_velocity(self, angularVelocity):
        return self.run_native("set_angular_velocity", angularVelocity)
    
    # _____________________________________________________________________
    
    def get_friction(self):
        return self.run_native("get_friction")
    
    # _____________________________________________________________________
    
    def set_friction(self, friction):
        return self.run_native("set_friction", friction)

    # _____________________________________________________________________
    
    def get_linear_damping(self):
        return self.run_native("get_linear_damping")
    
    # _____________________________________________________________________
    
    def set_linear_damping(self, linearDamping):
        return self.run_native("set_linear_damping", linearDamping)
    # _____________________________________________________________________
    
    def apply_torque(self, torque):
        return self.run_native("apply_torque", torque)
    # _____________________________________________________________________
    
    def apply_angular_impulse(self, impulse):
        return self.run_native("apply_angular_impulse", impulse)

    # _____________________________________________________________________
    
    def apply_force(self, force, point):
        fp_core_force = libfp_core.Point()
        fp_core_force.x = force.x
        fp_core_force.y = force.y
        
        fp_core_point = libfp_core.Point()
        fp_core_point.x = point.x
        fp_core_point.y = point.y
        
        return self.run_native("apply_force", fp_core_force, fp_core_point)
    
    # _____________________________________________________________________
    
    def set_is_sensor(self, isSensor):
        return self.run_native("set_is_sensor", isSensor)
    
    # _____________________________________________________________________
    
    def get_is_sensor(self):
        return self.run_native("get_is_sensor")
    # _____________________________________________________________________
    
    # ---------------------------------------------------------------------
    # Properties
    # ---------------------------------------------------------------------

    transformable = property(get_transformable, set_transformable)
    world = property(get_world, set_world)
    type = property(get_type, set_type)
    shape_scale = property(get_shape_scale, set_shape_scale)
    linear_velocity = property(get_linear_velocity, set_linear_velocity)
    angular_velocity = property(get_angular_velocity, set_angular_velocity)
    friction = property(get_friction, set_friction)
    linear_damping = property(get_linear_damping, set_linear_damping)
    isSensor = property(get_is_sensor, set_is_sensor)

NativeClass.class_map[libfp_core.Body] = Body

File: fp_core/src/main/src/Physics/Body.cpp
/*
 * Body.cpp
 *
 *  Created on: Sep 9, 2012
 *      Author: leo
 */

#include <fruitpunch/Physics/Body.h>
#include <fruitpunch/Graphics/Transformable.h>
#include <fruitpunch/Common/Runtime.h>
#include <fruitpunch/Physics/PhysicsWorld.h>
#include <fruitpunch/Errors/InvalidParentError.h>
#include <fruitpunch/Physics/PhysicsShape.h>
#include <typeinfo>

float PI180 = 0.01745329251f;

using namespace fp_core;
using namespace fp_core::common;

namespace fp_core {
namespace physics {

Body::Body() : Element(),
	mBody(NULL),
	mFixture(NULL) {
    mWorld = PhysicsWorld::getCurrent();
    mDensity = 1.0f;
    mFriction = 0.3f;
    mRestitution = 0.5f;
    mType = DYNAMIC;
    mShapeScale = 1.0f;

    if (mWorld) {
    	mIsSensor = mWorld->getBodiesSensorByDefault();
    }
}

// ____________________________________________________________________________

Body::~Body() {
    onRemove();
}

// ____________________________________________________________________________

void Body::create() {
  // ---------------------------------------------------------------------------
  // Checks if we are adding this body to an element
  // ---------------------------------------------------------------------------
  mTransformable = boost::dynamic_pointer_cast<Transformable>(getParent());
  if (!mTransformable) {
	 Runtime::getCurrent()->addError(InvalidParentError::ref(new InvalidParentError("Body", "Transformable")));
	 return;
  }


  // ---------------------------------------------------------------------------
  // Checks if this body exists. If so, destroy the old copy
  // ---------------------------------------------------------------------------
  if (mBody != NULL) {
    getWorld()->getWorld()->DestroyBody(mBody);
  }

  // ---------------------------------------------------------------------------
  // Creates the box2d body definition
  // ---------------------------------------------------------------------------
  mBodyDef = boost::shared_ptr<b2BodyDef>(new b2BodyDef());
  updateBodyDef();

  // ---------------------------------------------------------------------------
  // Creates the box2d fixture definition
  // ---------------------------------------------------------------------------
  mFixtureDef = boost::shared_ptr<b2FixtureDef>(new b2FixtureDef());

  // If the user has already defined a shape, then use that
  if (!mShape) {
	  mShape = getWorld()->makeShapeFor(getParent());

	  mBox2dShape = mShape->getBox2dShape(mShapeScale);
  }

  // if we still have a null shape, it's because the Shaper could not find a
  // valid one
  if (!mShape) {
    // TODO add proper runtime error handling
    printf("No shape generated!\n");
    return;
  }

  updateFixtureDef();

  // ---------------------------------------------------------------------------
  // Creates the box2d body
  // ---------------------------------------------------------------------------
  mBody = getWorld()->getWorld()->CreateBody(mBodyDef.get());
  mFixture = mBody->CreateFixture(mFixtureDef.get());

  // ---------------------------------------------------------------------------
  // Maps the fixture to the element in the PhysicsWorld so we can retrieve
  // in the contact event later.
  // ---------------------------------------------------------------------------
  getWorld()->addFixtureToElementMapping(getParent(), mFixture);
}

// ____________________________________________________________________________

void Body::updateBodyDef() {

  // ---------------------------------------------------------------------------
  // Define body type
  // ---------------------------------------------------------------------------
  switch (mType) {
  case (DYNAMIC):
      mBodyDef->type = b2_dynamicBody;
  break;
  case (STATIC):
      mBodyDef->type = b2_staticBody;
  break;
  case (KINEMATIC):
      mBodyDef->type = b2_kinematicBody;
  break;
  }

  // ---------------------------------------------------------------------------
  // Define body position and angle
  // ---------------------------------------------------------------------------
  float pixelSize = getWorld()->getPixelSize();
  mBodyDef->position.Set(mTransformable->getPosition().x / pixelSize,
      mTransformable->getPosition().y / pixelSize );
  mBodyDef->angle = mTransformable->getRotation() * PI180;
}

// ____________________________________________________________________________

void Body::renderFrame() {
  b2Vec2 position = mBody->GetPosition();
  point newPosition;
  newPosition.x = position.x * getWorld()->getPixelSize();
  newPosition.y = position.y * getWorld()->getPixelSize();
  getTransformable()->setPosition(newPosition);
  getTransformable()->setRotation(mBody->GetAngle() / PI180);

}

// ____________________________________________________________________________

void Body::onRemove() {

	if (mBody != NULL) {
		getWorld()->removeFixtureToElementMapping(mFixture);
		getWorld()->getWorld()->DestroyBody(mBody);
		mBody = NULL;
		mFixture = NULL;
	}

}

// ____________________________________________________________________________

void Body::onAdd() {
	create();
}

// ____________________________________________________________________________

float Body::getShapeScale() const {
	return mShapeScale;
}

void Body::setShapeScale(float shapeScale) {
	mShapeScale = shapeScale;
}

point Body::getLinearVelocity() const {
	return mLinearVelocity;
}

void Body::setLinearVelocity(point velocity) {
	mLinearVelocity = velocity;
	b2Vec2 vel;
	vel.x = velocity.x;
	vel.y = velocity.y;
	mBody->SetAwake(true);
	mBody->SetLinearVelocity(vel);
}

const float Body::getAngularVelocity() const {
	return mAngularVelocity;
}

void Body::setAngularVelocity(float angularVelocity) {
	mAngularVelocity = angularVelocity;
	mBody->SetAwake(true);
	mBody->SetAngularVelocity(angularVelocity);
}

void Body::updateFixtureDef() {
  mFixtureDef->density = mDensity;
  mFixtureDef->restitution = mRestitution;
  mFixtureDef->friction = mFriction;
  mFixtureDef->isSensor = mIsSensor;
  mFixtureDef->shape = mBox2dShape.get();
}

// ____________________________________________________________________________

Body::BodyType Body::getType() const {
  return mType;
}

// ---------------------------------------------------------------------------
// Getters and Setters
// ---------------------------------------------------------------------------

void Body::setType(Body::BodyType type) {
  mType = type;
}

const boost::shared_ptr<b2BodyDef>& Body::getBodyDef() const {
  return mBodyDef;
}

void Body::setBodyDef(const boost::shared_ptr<b2BodyDef>& bodyDef) {
  mBodyDef = bodyDef;
}

const boost::shared_ptr<b2FixtureDef>& Body::getFixtureDef() const {
  return mFixtureDef;
}

void Body::setFixtureDef(const boost::shared_ptr<b2FixtureDef>& fixtureDef) {
  mFixtureDef = fixtureDef;
}

const boost::shared_ptr<PhysicsShape>& Body::getShape() const {
  return mShape;
}

void Body::setShape(const boost::shared_ptr<PhysicsShape>& shape) {
  mShape = shape;
}

const boost::shared_ptr<Transformable>& Body::getTransformable() const {
  return mTransformable;
}

void Body::setTransformable(
    const boost::shared_ptr<Transformable>& transformable) {
  mTransformable = transformable;
}

boost::shared_ptr<PhysicsWorld> Body::getWorld() const {
  return mWorld;
}

void Body::setWorld(const boost::shared_ptr<PhysicsWorld>& world) {
  mWorld = world;
}

b2Body * Body::getBox2dBody() {
	return mBody;
}

void Body::setBox2dBody(b2Body * Box2dBody) {
	mBody = Box2dBody;
}

float Body::getFriction() {
	return mFriction;
}

void Body::setFriction(float friction) {
	mFriction = friction;
}

float Body::getLinearDamping() {
	return mLinearDamping;
}

void Body::setLinearDamping(float linearDamping) {
	mLinearDamping = linearDamping;
	mBody->SetLinearDamping(mLinearDamping);
}

void Body::applyTorque(float torque) {
	mBody->ApplyTorque(torque);
}

void Body::applyAngularImpulse(float impulse) {
	mBody->ApplyAngularImpulse(impulse);
}

void Body::applyForce(point force, point point) {
	b2Vec2 b2Force;
	b2Force.x = force.x;
	b2Force.y = force.y;

	b2Vec2 b2Point;
	b2Point.x = point.x;
	b2Point.y = point.y;
	mBody->ApplyForce(b2Force, b2Point);
}

bool Body::getIsSensor() const {
	return mIsSensor;
}

void Body::setIsSensor(bool isSensor) {
	mIsSensor = isSensor;
	mFixture->SetSensor(isSensor);
}

} /* namespace physics */
} /* namespace fp_core */
